### PR TITLE
ref(ts): Convert Share Issue to typescript

### PR DESCRIPTION
**Pull request by scttcper**
Was converting this to typescript and it had a few things that needed updating.

❌ Removed `FlowLayout`
❌ Inline styles -> ✅ Emotion + theme colors
❌ Removed duplicate busy/loading state inside share component
➕ Storybook stories

Did not change it from sharing automatically on click. Should look and work the same.

image

### CODE REVIEW

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+type ContainerProps = {
+  shareUrl: string;
+  onConfirming: () => void;
+  onConfirm: () => void;
+  onCancel: () => void;
+};
+
+class ShareUrlContainer extends React.Component<ContainerProps> {
+  urlRef?: AutoSelectText | null;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:25]: you are using this type in two different places, maybe you could create a common type and just reuse :)

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  /**
+   * Link is public
+   */
+  isShared?: boolean;
+  shareUrl?: string | null;
+  onToggle: () => void;
+  /**
+   * Called when refreshing an existing link
+   */
+  onReshare: () => void;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:88]:
```suggestion
  loading: boolean;
  /**
   * Called when refreshing an existing link
   */
  onReshare: () => void;
    /**
   * Link is public
   */
  onToggle: () => void;
  isShared?: boolean;
  shareUrl?: string | null;
```

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      onToggle();
+    }
+  };
+
+  // State of confirm modal so we can keep dropdown menu opn
+  handleConfirmCancel = () => (this.hasConfirmModal = false);
+  handleConfirmReshare = () => (this.hasConfirmModal = true);
+
+  render() {
+    const {loading, isShared = false, shareUrl, onReshare} = this.props;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:120]: in a class component, we usually declare the default value of a parameter in `defaultProps`

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+          <DropdownTitleContent>
+            <IndicatorDot isShared={isShared} />
+            {t('Share')}
+          </DropdownTitleContent>
+        }
+        onOpen={this.handleOpen}
+        keepMenuOpen
+      >
+        <DropdownContent>
+          <SpreadLayout style={{marginBottom: loading || isShared ? 12 : undefined}}>
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:136]: maybe you could use emotion to style this component?

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+  background: ${p => (p.isShared ? p.theme.green400 : p.theme.borderDark)};
+`;
+
+const StyledAutoSelectText = styled(AutoSelectText)<
+  React.ComponentPropsWithRef<typeof AutoSelectText>
+>`
+  flex: 1;
+  border: none;
+  padding: 4px 6px 4px 10px;
+  text-overflow: ellipsis;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:206]: maybe you could use the utility `overflowEllipsis` here

File: docs-ui/components/shareIssue.stories.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
+// eslint-disable-next-line sentry/no-react-hooks
+import React, {useState} from 'react';
```
Comment (priscilawebdev) [inline docs-ui/components/shareIssue.stories.js:2]: can we already use hooks? 😃👏

File: docs-ui/components/autoSelectText.stories.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import {withInfo} from '@storybook/addon-info';
+
+import AutoSelectText from 'app/components/autoSelectText';
+
+export default {
+  title: 'Utilities/AutoSelectText',
+};
+
+export const Default = withInfo('Select text on click')(() => (
+  <div>
```
Comment (priscilawebdev) [inline docs-ui/components/autoSelectText.stories.js:11]: do you need this `div` here?

File: docs-ui/components/shareIssue.stories.js (hunk 6 above)
Comment (markstory) [reply docs-ui/components/shareIssue.stories.js:2]: We have been avoiding them in the application code so far. I don't think we made decisions about what to do in non-application code.

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        <Confirm
+          message={t(
+            'You are about to regenerate a new shared URL. Your previously shared URL will no longer work. Do you want to continue?'
+          )}
+          onCancel={onCancel}
+          onConfirming={onConfirming}
+          onConfirm={onConfirm}
+        >
+          <RegenerateButton title={t('Regenerate URL')} borderless size="xsmall">
+            <IconRefresh />
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:69]: Should you use the icon prop of Button instead?

File: docs-ui/components/shareIssue.stories.js (hunk 6 above)
Comment (scttcper) [reply docs-ui/components/shareIssue.stories.js:2]: This was easier to write as a hook inside storybook for the example

Comment (scttcper) [reply docs-ui/components/shareIssue.stories.js]: Removed the hooks

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+const StyledAutoSelectText = styled(AutoSelectText)<
+  React.ComponentPropsWithRef<typeof AutoSelectText>
+>`
+  flex: 1;
+  padding: ${space(0.5)} 0 ${space(0.5)} ${space(0.75)};
+  ${overflowEllipsis}
+`;
+
+const StyledTextContainer = styled('div')`
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:221]: as you are not styling another component here, I'd name it only `TextContainer`  😉

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+  handleOpen = () => {
+    const {loading, isShared, onToggle} = this.props;
+    if (!loading && !isShared) {
+      // Starts sharing as soon as dropdown is opened
+      onToggle();
+    }
+  };
+
+  // State of confirm modal so we can keep dropdown menu opn
+  handleConfirmCancel = () => (this.hasConfirmModal = false);
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:120]:
```suggestion
  handleConfirmCancel = () => { this.hasConfirmModal = false };
```

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+  handleConfirmReshare = () => (this.hasConfirmModal = true);
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:121]: I think we should not return this 

```suggestion
  handleConfirmReshare = () => { this.hasConfirmModal = true };
```

File: src/sentry/static/sentry/app/components/shareIssue.tsx [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+          )}
+          onCancel={onCancel}
+          onConfirming={onConfirming}
+          onConfirm={onConfirm}
+        >
+          <ReshareButton
+            title={t('Generate new URL')}
+            borderless
+            size="xsmall"
+            icon={<IconRefresh />}
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/shareIssue.tsx:73]: 👏

File: docs-ui/components/shareIssue.stories.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import {text} from '@storybook/addon-knobs';
+import {withInfo} from '@storybook/addon-info';
+
+import ShareIssue from 'app/components/shareIssue';
+
+export default {
+  title: 'Features/Issues/Share Issue',
+};
+
+class ShareSimulator extends React.Component {
```
Comment (priscilawebdev) [inline docs-ui/components/shareIssue.stories.js:12]: this could be a function component ..but it's a minor thing

**Review by priscilawebdev — APPROVED** [review]
I left some more suggestions, but they should not block you...good job! 👏

**Pull request merged**